### PR TITLE
Adding mapping to declarative plotting

### PR DESCRIPTION
**Pull request by kevindougherty-noaa**

### CODE REVIEW

File: src/plots/declarative.py [hunk 1]
```diff
@@ -27,21 +32,63 @@ def create_figure(self):
...
+            'map_scatter': self._map_scatter,
+            'map_gridded': self._map_gridded,
+            'map_contour': self._map_contour
         }
-        # Feature dictionary
+
+        gs = gridspec.GridSpec(self.nrows, self.ncols)
+        self.fig = plt.figure(figsize=self.figsize)
+
+        for i in range(0, self.nrows*self.ncols):
```
Comment (CoryMartin-NOAA) [inline src/plots/declarative.py]: for i, plotobj in enumerate(self.plot_list): or whatever the syntax is?

File: src/plots/declarative.py [hunk 2]
```diff
@@ -59,30 +106,61 @@ def create_figure(self):
...
+                            s=plotobj.markersize, **inputs,
+                            transform=self.projection.projection)
+        else:
+            skipvars = ['plottype', 'longitude', 'latitude',
+                        'data', 'markersize']
+            inputs = self._get_inputs_dict(skipvars, plotobj)
+            cs = ax.scatter(plotobj.longitude, plotobj.latitude,
+                            c=plotobj.data, s=plotobj.markersize,
+                            **inputs, transform=self.projection.projection)
+#         if plotobj.colorbar:
```
Comment (CoryMartin-NOAA) [inline src/plots/declarative.py:141]: comments?

Comment (CoryMartin-NOAA) [reply src/plots/declarative.py:141]: I guess colorbar isn't working yet, did I miss that?

File: src/plots/declarative.py [hunk 3]
```diff
@@ -361,21 +451,110 @@ def _set_yscale(self, ax, yscale):
         """
         ax.set_yscale(yscale)
 
+    def _sharex(self, ax):
+        """
+        If sharex axis is True, will find where to hide xticklabels.
+        """
+        if ax.is_last_row():
```
Comment (CoryMartin-NOAA) [inline src/plots/declarative.py]: if this returns True, do `if not ax.is_last_row(): plt.setp...`

File: src/plots/declarative.py [hunk 4]
```diff
@@ -361,21 +451,110 @@ def _set_yscale(self, ax, yscale):
...
+            pass
+        else:
+            plt.setp(ax.get_xticklabels(), visible=False)
+
+    def _sharey(self, ax):
+        """
+        If sharey axis is True, will find where to hide yticklabels.
+        """
+        if ax.is_first_col():
```
Comment (CoryMartin-NOAA) [inline src/plots/declarative.py]: Same comment

File: src/plots/declarative.py [hunk 5]
```diff
@@ -361,21 +451,110 @@ def _set_yscale(self, ax, yscale):
...
+
+        for feat in map_features:
+            try:
+                feature_dict[feat](ax)
+            except KeyError:
+                raise TypeError(f'{feat} is not a valid map feature.' +
+                                'Current map features supported are:\n' +
+                                f'{" | ".join(feature_dict.keys())}"')
+
+    def _add_coastlines(self, ax):
```
Comment (CoryMartin-NOAA) [inline src/plots/declarative.py]: Can we make this into a factory such that you just pass the add_feature(cfeature.COASTLINE)?

Comment (CoryMartin-NOAA) [reply src/plots/declarative.py]: Meaning in your dict above, put `cfeature.COASTLINE` and then just one `def _add_feature`

Comment (kevindougherty-noaa) [reply src/plots/declarative.py]: I'll do you one up: 
```
feature_dict = {
    'coastline': cfeature.COASTLINE,
    'borders': cfeature.BORDERS,
    'states': cfeature.STATES,
    'lakes': cfeature.LAKES,
    'rivers': cfeature.RIVERS,
    'land': cfeature.LAND,
    'ocean': cfeature.OCEAN
}

for feat in map_features:
    try:
        ax.add_feature(feature_dict[feat])
    except:
        ....
```

File: src/plots/declarative.py (hunk 2 above)
Comment (kevindougherty-noaa) [reply src/plots/declarative.py:141]: Yes, still working out the kinks of colorbar. It gets tricky, I will continue to work on this tomorrow.

File: src/plots/declarative.py (hunk 1 above)
Comment (kevindougherty-noaa) [reply src/plots/declarative.py]: Great suggestion., thanks.

File: src/plots/declarative.py [hunk 6]
```diff
@@ -361,21 +459,64 @@ def _set_yscale(self, ax, yscale):
...
+                 domain=None):
 
         self.plot_layers = plot_layers
 
+        ###############################################
+        # Need a better way of doing this
+        if projection is not None and domain is not None:
+            self.projection = projection
+            self.domain = domain
+        ###############################################
```
Comment (kevindougherty-noaa) [inline src/plots/declarative.py:514]: @CoryMartin-NOAA what are your thoughts on this? It is essential for determining if the object in the loop in `create_figure()` is a map or standard plot. Is there a better way?

Comment (CoryMartin-NOAA) [reply src/plots/declarative.py:514]: I think it's okay for now

**Review by CoryMartin-NOAA — APPROVED** [review]

**Pull request merged**